### DISCUSSION TITLE
Node app, Prisma, pkg

### DISCUSSION THREAD

**Discussion by jjjuk** [prisma/prisma#6892] [Q&A] [answered]
Hello everyone! 
I'm trying to pack my node app into executable with pkg. But It instantly crashes when I try to run it. 
Here's my package.json:
```json

 "bin": "build/index.js",
  "pkg": {
    "scripts": "build/**/*.js",
    "assets": [
      "prisma/schema.prisma",
      "prisma/local.db",
      "node_modules/.prisma/**/*"
    ],
    "targets": [
      "node14-linux-x64",
      "node14-macos-x64",
      "node14-win-x64"
    ],
    "outputPath": "dist"
  }

```
Could someone help me? Is there some proper way to do it? Or it's imposable because Prisma uses Rust and communicate with another process and all that...(

**Comment by janpio** [comment, marked as answer]
FINALLY!

Finally someone is trying to do this. (cc @steebchen)

We have e2e tests for this at https://github.com/prisma/e2e-tests/tree/dev/binaries/pkg which show how this works. The Windows version is pretty ugly, but could/should work.

**Reply by marcelherd** [reply]
Do you also support the combination of pkg + prisma with sqlite?

I specified the datasource in my schema like so:

```
datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}
```

And run the pkg created .exe like so:

```
cmd /V /C "set DATABASE_URL=file:./database.db&& start app.exe"
```

Where `app.exe` and `database.db` are both in the same folder. However, when running app.exe, it cannot find the database file:

```
PrismaClientInitializationError:
Invalid `prisma.match.groupBy()` invocation:

  Error querying the database: unable to open database file: C:\snapshot\tlp-webapp\node_modules\.prisma\client\./database.db
    at Object.request (C:\snapshot\tlp-webapp\node_modules\@prisma\client\runtime\index.js:45411:15)
    at async PrismaClient._request (C:\snapshot\tlp-webapp\node_modules\@prisma\client\runtime\index.js:46301:18)
    at async handler (C:\snapshot\tlp-webapp\.next\server\pages\api\overview.js:56:30)
    at async Object.apiResolver (C:\snapshot\tlp-webapp\node_modules\next\dist\server\api-utils\node.js:185:9)
    at async NextNodeServer.runApi (C:\snapshot\tlp-webapp\node_modules\next\dist\server\next-server.js:395:9)
    at async Object.fn (C:\snapshot\tlp-webapp\node_modules\next\dist\server\base-server.js:496:37)
    at async Router.execute (C:\snapshot\tlp-webapp\node_modules\next\dist\server\router.js:226:36)
    at async NextNodeServer.run (C:\snapshot\tlp-webapp\node_modules\next\dist\server\base-server.js:606:29)
    at async NextNodeServer.handleRequest (C:\snapshot\tlp-webapp\node_modules\next\dist\server\base-server.js:321:20) {
  clientVersion: '3.14.0',
  errorCode: undefined
}
```

Am I doing anything wrong or is this just not supported? I cannot include the database file during build

**Reply by marcelherd** [reply]
I found a working method. I define the datasource URL like so when instantiating my Prisma Client:

```typescript
new PrismaClient({
    log: ['query'],
    datasources: {
      db: {
        url: `file:${path.join(process.cwd(), 'database.db')}`,
      },
    },
  });
```

**Reply by janpio** [reply]
I think the difference between the two might be that the code version sets the full path, while the env var tries to set the relative one? Did you try the end var with the full, expanded path as well?

**Comment by jjjuk** [comment]
> Finally someone is trying to do this. (cc @steebchen)

@janpio OMG! You've been waiting! 👀 

Still, It crashes, I have no ideas, I'm not good in this things(
It's cool that prisma builds in pkg, but how should I build my app with prisma as dependency?

> We have e2e tests for this at https://github.com/prisma/e2e-tests/tree/dev/binaries/pkg

This one doesn't help me here

I have this warning:

![image](https://user-images.githubusercontent.com/61079830/116950839-ddcb9a00-acb8-11eb-96b5-b6a7ae8eafac.png)

Seems simple. But where should I put this binaries, what kind of assets I should specify in package.json. Do I need schema? probably not.. And I use sqlite so I think I need asset for db. 🤕 🤕 🤕  I need help(

**Reply by steebchen** [reply]
Is this just a warning? Or is this the actual error which prevents building the binaries?

**Reply by jjjuk** [reply]
> Is this just a warning? Or is this the actual error which prevents building the binaries?

Build still crashes immediately, nothing works

**Reply by jjjuk** [reply]
Could you make a little reproduction with simple server app that works in pkg binaries and queries something from sqlite like: 
```js
app.get('/', async (_, res) => {  
const somethings = await prisma.something.findMany()
res.send(JSON.stringify({ data: somethings })) 
})
```
please?

**Reply by steebchen** [reply]
I can't help you without more information.

Try this:

```
yarn init
yarn add pkg -D
yarn add prisma -D
yarn add @prisma/client

# special commands
mkdir -p node_modules/prisma/node_modules/@prisma/engines
cp -R node_modules/@prisma/engines/* node_modules/prisma/node_modules/@prisma/engines

yarn pkg node_modules/prisma -t node12-macos # or whatever your operating system is

# and then finally run it
./prisma
```

Please let me know which command throws an error, and if it does, what exactly the error is.


If it does work, I'd suggest to try running the "special commands" above in _your_ project, then re-run `pkg` (best using the same command/config as I did above) and check if it works.

**Reply by jjjuk** [reply]
Prisma exe works fine with node14-windows!

![image](https://user-images.githubusercontent.com/61079830/117063686-be377e80-ad57-11eb-8d07-3e1c21514ce9.png)

But still build of the whole app throws errors and crushes
![image](https://user-images.githubusercontent.com/61079830/117063881-0787ce00-ad58-11eb-84e3-da068e2df5e9.png)
![image](https://user-images.githubusercontent.com/61079830/117063916-11a9cc80-ad58-11eb-9a8c-4972bc980d89.png)

Any assets don't make a difference

**Reply by janpio** [reply]
The warnings usually are not a problem.
The Error you are getting though is, but that seems to be a problem in your application, some dependency you are using. Does that work if you are not including Prisma in your application?

**Reply by jjjuk** [reply]
Yeah, I've just realized that I need to check all these. I think I'll come back here soon!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
